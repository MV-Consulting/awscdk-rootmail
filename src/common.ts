const integTestWildcard = '${Token[AWS.Region';
// see https://aws.amazon.com/de/about-aws/whats-new/2023/09/amazon-ses-email-service-7-regions/
export const sesEnabledRegions = [
  'us-east-1', // US East (N. Virginia),
  'eu-west-1', // Europe (Ireland),
  'us-west-2', // US West (Oregon),
  // new regions since 2023-09
  'eu-central-1', // Europe (Frankfurt),
  'eu-west-2', // Europe (London),
  'us-east-2', // US East (Ohio),
  'ca-central-1', // Canada (Central),
  'ap-northeast-1', // Asia Pacific (Tokyo),
  'ap-southeast-1', // Asia Pacific (Singapore),
  'ap-southeast-2', // Asia Pacific (Sydney),
];

export function isSESEnabledRegion(region: string): boolean {
  return sesEnabledRegions.includes(region) || region.startsWith(integTestWildcard);
}