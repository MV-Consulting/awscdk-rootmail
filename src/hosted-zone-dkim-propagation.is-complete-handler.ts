import { SES } from '@aws-sdk/client-ses';
// eslint-disable-next-line import/no-unresolved
import * as AWSCDKAsyncCustomResource from 'aws-cdk-lib/custom-resources/lib/provider-framework/types';
export const PROP_DOMAIN = 'Domain';

const ses = new SES();

export interface IsCompleteHandlerResponse {
  IsComplete: boolean;
}

export async function handler(event: AWSCDKAsyncCustomResource.OnEventRequest): Promise<IsCompleteHandlerResponse> {
  const domain = event.ResourceProperties[PROP_DOMAIN];

  switch (event.RequestType) {
    case 'Create':
      const isReady = await internalHandler(domain);
      return { IsComplete: isReady };
    case 'Update':
    case 'Delete':
      return {
        IsComplete: true,
      };
  }
}

function log(msg: any) {
  console.log(JSON.stringify(msg));
}

async function internalHandler(domain: string): Promise<boolean> {

  log({
    domain: domain,
    level: 'debug',
  });

  const sendingResponse = await ses.getAccountSendingEnabled({});
  if (!sendingResponse.Enabled) {
    return false;
  }
  log('sending enabled');

  const identityVerificationResponse = await ses.getIdentityVerificationAttributes({ Identities: [domain] });
  const identityVerificationStatus = identityVerificationResponse.VerificationAttributes![domain].VerificationStatus;
  if (identityVerificationStatus !== 'Success') {
    log(`Identity Verification status not successful. Was '${identityVerificationStatus}'`);
    return false;
  }
  log('identitity verification successful');

  const identityDkimRes = await ses.getIdentityDkimAttributes({ Identities: [domain] });
  const identityDkimStatus = identityDkimRes.DkimAttributes![domain].DkimVerificationStatus;
  if (identityDkimStatus !== 'Success') {
    log(`DKIM status not successful. Was '${identityDkimStatus}'`);
    return false;
  }
  log('DKIM verification successful');

  const identityNotificationRes = await ses.getIdentityNotificationAttributes({ Identities: [domain] });
  const forwardingEnabled = identityNotificationRes.NotificationAttributes![domain].ForwardingEnabled;
  if (!forwardingEnabled) {
    log(`Forwarding not enabled. Was '${forwardingEnabled}'`);
    return false;
  }
  log('forwarding enabled');
  return true;
}

