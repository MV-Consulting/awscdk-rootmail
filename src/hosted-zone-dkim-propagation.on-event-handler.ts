// eslint-disable-next-line import/no-unresolved
import * as AWSCDKAsyncCustomResource from 'aws-cdk-lib/custom-resources/lib/provider-framework/types';
export const PROP_DOMAIN = 'Domain';

export async function handler(event: AWSCDKAsyncCustomResource.OnEventRequest): Promise<AWSCDKAsyncCustomResource.OnEventResponse> {
  switch (event.RequestType) {
    case 'Create':
    case 'Update':
    case 'Delete':
      console.log(`${event.RequestType} DKIM propagation, doing nothing`);
      return {
        PhysicalResourceId: event.PhysicalResourceId,
      };
  }
}

