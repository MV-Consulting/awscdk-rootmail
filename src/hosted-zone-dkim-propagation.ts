import {
  CustomResource,
  Duration,
  Stack,
  aws_iam as iam,
  aws_lambda as lambda,
} from 'aws-cdk-lib';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import * as cr from 'aws-cdk-lib/custom-resources';
import { NagSuppressions } from 'cdk-nag';
import { Construct, Node } from 'constructs';
import { PROP_DOMAIN } from './hosted-zone-dkim-propagation.on-event-handler';

export interface HostedZoneDKIMPropagationProps {
  readonly domain: string;
  readonly totalTimeToWireDNS?: Duration;
}

export class HostedZoneDKIMPropagation extends Construct {
  constructor(scope: Construct, id: string, props: HostedZoneDKIMPropagationProps) {
    super(scope, id);

    // TODO can we pass in custom cloudwatch log groups? so they are deleted on destroy

    new CustomResource(this, 'Resource', {
      serviceToken: HostedZoneDKIMPropagationProvider.getOrCreate(this, { totalTimeToWireDNS: props.totalTimeToWireDNS }),
      resourceType: 'Custom::HostedZoneDKIMPropagation',
      properties: {
        [PROP_DOMAIN]: props.domain,
      },
    });
  }
}

interface HostedZoneDKIMPropagationProviderProps {
  readonly totalTimeToWireDNS?: Duration;
}

class HostedZoneDKIMPropagationProvider extends Construct {

  /**
   * Returns the singleton provider.
   */
  public static getOrCreate(scope: Construct, props: HostedZoneDKIMPropagationProviderProps) {
    const stack = Stack.of(scope);
    const id = 'rootmail.hosted-zone-dkim-propagation-provider';
    const x = Node.of(stack).tryFindChild(id) as HostedZoneDKIMPropagationProvider
      || new HostedZoneDKIMPropagationProvider(stack, id, props);
    return x.provider.serviceToken;
  }

  private readonly provider: cr.Provider;

  constructor(scope: Construct, id: string, props: HostedZoneDKIMPropagationProviderProps) {
    super(scope, id);

    const isCompleteHandlerFunc = new NodejsFunction(this, 'is-complete-handler', {
      runtime: lambda.Runtime.NODEJS_18_X,
      logRetention: 3,
      timeout: Duration.seconds(30),
      bundling: {
        esbuildArgs: {
          '--packages': 'bundle',
        },
      },
    });

    isCompleteHandlerFunc.addToRolePolicy(
      new iam.PolicyStatement({
        actions: [
          'ses:GetIdentityVerificationAttributes',
          'ses:GetAccountSendingEnabled',
          'ses:GetIdentityDkimAttributes',
          'ses:GetIdentityNotificationAttributes',
        ],
        effect: iam.Effect.ALLOW,
        resources: ['*'],
      }),
    );
    NagSuppressions.addResourceSuppressions(isCompleteHandlerFunc, [
      { id: 'AwsSolutions-IAM4', reason: 'no service role restriction needed' },
      { id: 'AwsSolutions-IAM5', reason: 'wildcards are ok for the ses mail verification' },
    ], true);

    const onEventHandlerFunc = new NodejsFunction(this, 'on-event-handler', {
      runtime: lambda.Runtime.NODEJS_18_X,
      logRetention: 3,
      timeout: Duration.seconds(10),
      bundling: {
        esbuildArgs: {
          '--packages': 'bundle',
        },
      },
    });
    NagSuppressions.addResourceSuppressions(onEventHandlerFunc, [
      { id: 'AwsSolutions-IAM4', reason: 'no service role restriction needed' },
      { id: 'AwsSolutions-IAM5', reason: 'wildcards are ok as the function by itself does nothing' },
    ], true);

    this.provider = new cr.Provider(this, 'hosted-zone-dkim-propagation-provider', {
      isCompleteHandler: isCompleteHandlerFunc,
      queryInterval: Duration.seconds(10),
      totalTimeout: props.totalTimeToWireDNS,
      onEventHandler: onEventHandlerFunc,
    });
    NagSuppressions.addResourceSuppressions(this.provider, [
      { id: 'AwsSolutions-IAM4', reason: 'no service role restriction needed' },
      { id: 'AwsSolutions-IAM5', reason: 'wildcards are ok for the provider as the function has restrictions' },
    ], true);
  };
}

