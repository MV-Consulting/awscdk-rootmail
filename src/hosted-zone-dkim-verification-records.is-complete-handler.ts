// eslint-disable-next-line import/no-unresolved
import * as AWSCDKAsyncCustomResource from 'aws-cdk-lib/custom-resources/lib/provider-framework/types';
import * as AWS from 'aws-sdk';
export const PROP_DOMAIN = 'Domain';
export const ATTR_VERIFICATION_TOKEN = 'VerificationToken';
export const ATTR_DKIM_TOKENS = 'DkimTokens';

// this is fixed to eu-west-1 until SES supports receive more globally (see #23)
const SES = new AWS.SES({ region: 'eu-west-1' });

export interface IsCompleteHandlerResponse {
  IsComplete: boolean;
}

export async function handler(event: AWSCDKAsyncCustomResource.OnEventRequest): Promise<IsCompleteHandlerResponse> {
  const domain = event.ResourceProperties[PROP_DOMAIN];
  let isReady = true;

  console.log(domain);
  console.log(SES);

  return {
    IsComplete: isReady,
  };
}

