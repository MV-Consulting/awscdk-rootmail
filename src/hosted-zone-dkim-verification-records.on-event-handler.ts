// eslint-disable-next-line import/no-unresolved
import * as AWSCDKAsyncCustomResource from 'aws-cdk-lib/custom-resources/lib/provider-framework/types';
import * as AWS from 'aws-sdk';
export const PROP_DOMAIN = 'Domain';
export const ATTR_VERIFICATION_TOKEN = 'VerificationToken';
export const ATTR_DKIM_TOKENS = 'DkimTokens';

const SES = new AWS.SES();

export async function handler(event: AWSCDKAsyncCustomResource.OnEventRequest): Promise<AWSCDKAsyncCustomResource.OnEventResponse> {
  const domain = event.ResourceProperties[PROP_DOMAIN];
  switch (event.RequestType) {
    case 'Create':
    case 'Update': // TODO check for double creation
      let physicalResourceId = event.PhysicalResourceId;
      if (event.RequestType === 'Create') {
        physicalResourceId = event.RequestId;
      }

      console.log(`${event.RequestType}: Do Domain verification and DKIM records for ${event.LogicalResourceId} and domain '${domain}' with PhysicalResourceId '${physicalResourceId}'`);
      const verifyDomainResponse = await SES.verifyDomainIdentity({ Domain: domain }).promise();
      const verificationToken = verifyDomainResponse.VerificationToken;
      console.log(`${event.RequestType}: Got verification token '${verificationToken}' for domain '${domain}'`);

      const verifyDomainDkimResponse = await SES.verifyDomainDkim({ Domain: domain }).promise();
      const dkimTokens = verifyDomainDkimResponse.DkimTokens;
      console.log(`${event.RequestType}: Got DKIM tokens '${dkimTokens}' for domain '${domain}'`);


      return {
        PhysicalResourceId: physicalResourceId,
        Data: {
          [ATTR_VERIFICATION_TOKEN]: verificationToken,
          [ATTR_DKIM_TOKENS]: dkimTokens,
        },
      };
    case 'Delete':
      console.log(`Deleting Domain identity for domain '${domain}' with PhysicalResourceId '${event.PhysicalResourceId}'`);
      const deleteResponse = await SES.deleteIdentity({ Identity: domain }).promise();
      console.log(`Deleted Domain identity for domain '${domain}'`, deleteResponse);
      return {
        PhysicalResourceId: event.PhysicalResourceId,
      };
  }
}

