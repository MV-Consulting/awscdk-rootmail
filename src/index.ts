// TODO remove
export * from './ses-receive';
export * from './rootmail';