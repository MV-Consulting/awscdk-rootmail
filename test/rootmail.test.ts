import {
  App,
  Stack,
} from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { Rootmail, RootmailProps } from '../src';

test.skip('rootmail create', () => {
  const app = new App();
  const stack = new Stack(app, 'testStack', {
    env: {
      region: 'us-east-1',
      account: '1234',
    },
  });

  const testProps: RootmailProps = {
    domain: 'example.com',
  };

  new Rootmail(stack, 'testRootmail', testProps);

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});