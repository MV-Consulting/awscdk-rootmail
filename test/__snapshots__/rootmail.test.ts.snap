// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rootmail create 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "superwerkerhostedzonedkimverificationrecordsproviderframeworkonEvent0173C90D": {
      "DependsOn": [
        "superwerkerhostedzonedkimverificationrecordsproviderframeworkonEventServiceRoleDefaultPolicy4EBCA739",
        "superwerkerhostedzonedkimverificationrecordsproviderframeworkonEventServiceRoleFE1295B9",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-1234-us-east-1",
          "S3Key": "f2d30cfc360482320a52a4fcde8a70f3569df79ab30be24650fda58eb60052cf.zip",
        },
        "Description": "AWS CDK resource provider framework - onEvent (testStack/superwerker.hosted-zone-dkim-verification-records-provider/hosted-zone-dkim-verification-records-provider)",
        "Environment": {
          "Variables": {
            "USER_ON_EVENT_FUNCTION_ARN": {
              "Fn::GetAtt": [
                "superwerkerhostedzonedkimverificationrecordsproviderhostedzonedkimverificationrecordsonevent5D8C939A",
                "Arn",
              ],
            },
          },
        },
        "Handler": "framework.onEvent",
        "Role": {
          "Fn::GetAtt": [
            "superwerkerhostedzonedkimverificationrecordsproviderframeworkonEventServiceRoleFE1295B9",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "superwerkerhostedzonedkimverificationrecordsproviderframeworkonEventServiceRoleDefaultPolicy4EBCA739": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "superwerkerhostedzonedkimverificationrecordsproviderhostedzonedkimverificationrecordsonevent5D8C939A",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "superwerkerhostedzonedkimverificationrecordsproviderhostedzonedkimverificationrecordsonevent5D8C939A",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "superwerkerhostedzonedkimverificationrecordsproviderframeworkonEventServiceRoleDefaultPolicy4EBCA739",
        "Roles": [
          {
            "Ref": "superwerkerhostedzonedkimverificationrecordsproviderframeworkonEventServiceRoleFE1295B9",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "superwerkerhostedzonedkimverificationrecordsproviderframeworkonEventServiceRoleFE1295B9": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "superwerkerhostedzonedkimverificationrecordsproviderhostedzonedkimverificationrecordsonevent5D8C939A": {
      "DependsOn": [
        "superwerkerhostedzonedkimverificationrecordsproviderhostedzonedkimverificationrecordsoneventServiceRoleDefaultPolicyA2E96F86",
        "superwerkerhostedzonedkimverificationrecordsproviderhostedzonedkimverificationrecordsoneventServiceRoleE9B9349F",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-1234-us-east-1",
          "S3Key": "e8436f52df70c502857b6ab342eedcede241b255a290064de01c70dd49c663af.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "superwerkerhostedzonedkimverificationrecordsproviderhostedzonedkimverificationrecordsoneventServiceRoleE9B9349F",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 200,
      },
      "Type": "AWS::Lambda::Function",
    },
    "superwerkerhostedzonedkimverificationrecordsproviderhostedzonedkimverificationrecordsoneventServiceRoleDefaultPolicyA2E96F86": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "ses:VerifyDomainDkim",
                "ses:VerifyDomainIdentity",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "superwerkerhostedzonedkimverificationrecordsproviderhostedzonedkimverificationrecordsoneventServiceRoleDefaultPolicyA2E96F86",
        "Roles": [
          {
            "Ref": "superwerkerhostedzonedkimverificationrecordsproviderhostedzonedkimverificationrecordsoneventServiceRoleE9B9349F",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "superwerkerhostedzonedkimverificationrecordsproviderhostedzonedkimverificationrecordsoneventServiceRoleE9B9349F": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "testRootmailEmailBucket2A470446": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketName": "teststacktmailemailbuckete12ab3f3d5ea6a1fea86",
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "testRootmailEmailBucketPolicy5DA39984": {
      "Properties": {
        "Bucket": {
          "Ref": "testRootmailEmailBucket2A470446",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:PutObject",
              "Condition": {
                "StringEquals": {
                  "aws:Referer": "1234",
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "ses.amazonaws.com",
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":s3:::",
                    {
                      "Ref": "testRootmailEmailBucket2A470446",
                    },
                    "/RootMail/*",
                  ],
                ],
              },
              "Sid": "EnableSESReceive",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "testRootmailErrors174176D2": {
      "Properties": {
        "AlarmName": "superwerker-RootMailReady",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "FunctionName",
            "Value": {
              "Ref": "testRootmailRootMailReady825120E7",
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "Errors",
        "Namespace": "AWS/Lambda",
        "Period": 180,
        "Statistic": "Sum",
        "Threshold": 1,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "testRootmailHostedZone6DFD3C9C": {
      "Properties": {
        "Name": "aws.example.com.",
      },
      "Type": "AWS::Route53::HostedZone",
    },
    "testRootmailHostedZoneDKIMAndVerificationRecordsE48944E4": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "Domain": "example.com",
        "ServiceToken": {
          "Fn::GetAtt": [
            "superwerkerhostedzonedkimverificationrecordsproviderframeworkonEvent0173C90D",
            "Arn",
          ],
        },
      },
      "Type": "Custom::HostedZoneDKIMAndVerificationRecords",
      "UpdateReplacePolicy": "Delete",
    },
    "testRootmailHostedZoneDKIMTokenRecord0E194A84E": {
      "Properties": {
        "HostedZoneId": {
          "Ref": "testRootmailHostedZone6DFD3C9C",
        },
        "Name": {
          "Fn::Join": [
            "",
            [
              {
                "Fn::Select": [
                  0,
                  {
                    "Fn::GetAtt": [
                      "testRootmailHostedZoneDKIMAndVerificationRecordsE48944E4",
                      "DkimTokens",
                    ],
                  },
                ],
              },
              "_domainkey.aws.example.com.",
            ],
          ],
        },
        "ResourceRecords": [
          {
            "Fn::Join": [
              "",
              [
                {
                  "Fn::Select": [
                    0,
                    {
                      "Fn::GetAtt": [
                        "testRootmailHostedZoneDKIMAndVerificationRecordsE48944E4",
                        "DkimTokens",
                      ],
                    },
                  ],
                },
                ".dkim.amazonses.com",
              ],
            ],
          },
        ],
        "TTL": "60",
        "Type": "CNAME",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "testRootmailHostedZoneDKIMTokenRecord1F2425011": {
      "Properties": {
        "HostedZoneId": {
          "Ref": "testRootmailHostedZone6DFD3C9C",
        },
        "Name": {
          "Fn::Join": [
            "",
            [
              {
                "Fn::Select": [
                  1,
                  {
                    "Fn::GetAtt": [
                      "testRootmailHostedZoneDKIMAndVerificationRecordsE48944E4",
                      "DkimTokens",
                    ],
                  },
                ],
              },
              "_domainkey.aws.example.com.",
            ],
          ],
        },
        "ResourceRecords": [
          {
            "Fn::Join": [
              "",
              [
                {
                  "Fn::Select": [
                    1,
                    {
                      "Fn::GetAtt": [
                        "testRootmailHostedZoneDKIMAndVerificationRecordsE48944E4",
                        "DkimTokens",
                      ],
                    },
                  ],
                },
                ".dkim.amazonses.com",
              ],
            ],
          },
        ],
        "TTL": "60",
        "Type": "CNAME",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "testRootmailHostedZoneDKIMTokenRecord2F143C5BA": {
      "Properties": {
        "HostedZoneId": {
          "Ref": "testRootmailHostedZone6DFD3C9C",
        },
        "Name": {
          "Fn::Join": [
            "",
            [
              {
                "Fn::Select": [
                  2,
                  {
                    "Fn::GetAtt": [
                      "testRootmailHostedZoneDKIMAndVerificationRecordsE48944E4",
                      "DkimTokens",
                    ],
                  },
                ],
              },
              "_domainkey.aws.example.com.",
            ],
          ],
        },
        "ResourceRecords": [
          {
            "Fn::Join": [
              "",
              [
                {
                  "Fn::Select": [
                    2,
                    {
                      "Fn::GetAtt": [
                        "testRootmailHostedZoneDKIMAndVerificationRecordsE48944E4",
                        "DkimTokens",
                      ],
                    },
                  ],
                },
                ".dkim.amazonses.com",
              ],
            ],
          },
        ],
        "TTL": "60",
        "Type": "CNAME",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "testRootmailHostedZoneMXRecordA975A9BB": {
      "Properties": {
        "HostedZoneId": {
          "Ref": "testRootmailHostedZone6DFD3C9C",
        },
        "Name": "aws.example.com.",
        "ResourceRecords": [
          "10 inbound-smtp.eu-west-1.amazonaws.com",
        ],
        "TTL": "60",
        "Type": "MX",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "testRootmailHostedZoneSSMParameter574DA41C": {
      "Properties": {
        "Name": "/superwerker/domain_name_servers",
        "Type": "StringList",
        "Value": {
          "Fn::Join": [
            ",",
            {
              "Fn::GetAtt": [
                "testRootmailHostedZone6DFD3C9C",
                "NameServers",
              ],
            },
          ],
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "testRootmailHostedZoneVerificationTokenRecordB9EB1498": {
      "Properties": {
        "HostedZoneId": {
          "Ref": "testRootmailHostedZone6DFD3C9C",
        },
        "Name": "_amazonses.aws.example.com.",
        "ResourceRecords": [
          {
            "Fn::GetAtt": [
              "testRootmailHostedZoneDKIMAndVerificationRecordsE48944E4",
              "VerificationToken",
            ],
          },
        ],
        "TTL": "60",
        "Type": "TXT",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "testRootmailRootMailReady825120E7": {
      "DependsOn": [
        "testRootmailRootMailReadyServiceRoleDefaultPolicy9A1F5C41",
        "testRootmailRootMailReadyServiceRole28CD8F41",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-1234-us-east-1",
          "S3Key": "1e0595a2505f679e82e0e64868db3f4d6f83f1cbc38d3ef0e372560b88fa7489.zip",
        },
        "Environment": {
          "Variables": {
            "DOMAIN": "example.com",
            "SUB_DOMAIN": "aws",
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "testRootmailRootMailReadyServiceRole28CD8F41",
            "Arn",
          ],
        },
        "Runtime": "python3.9",
        "Timeout": 260,
      },
      "Type": "AWS::Lambda::Function",
    },
    "testRootmailRootMailReadyEventRule77EE7FF3": {
      "Properties": {
        "ScheduleExpression": "cron(5 * * * ? *)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "testRootmailRootMailReady825120E7",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "testRootmailRootMailReadyEventRuleAllowEventRuletestStacktestRootmailRootMailReady5BA57D95B365F7DA": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "testRootmailRootMailReady825120E7",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "testRootmailRootMailReadyEventRule77EE7FF3",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "testRootmailRootMailReadyHandleC6B1EB65": {
      "Type": "AWS::CloudFormation::WaitConditionHandle",
    },
    "testRootmailRootMailReadyHandleWaitConditionAE1C01F5": {
      "Properties": {
        "Handle": {
          "Ref": "testRootmailRootMailReadyHandleC6B1EB65",
        },
        "Timeout": "28800",
      },
      "Type": "AWS::CloudFormation::WaitCondition",
    },
    "testRootmailRootMailReadyServiceRole28CD8F41": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "testRootmailRootMailReadyServiceRoleDefaultPolicy9A1F5C41": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "ses:GetIdentityVerificationAttributes",
                "ses:GetAccountSendingEnabled",
                "ses:GetIdentityDkimAttributes",
                "ses:GetIdentityNotificationAttributes",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testRootmailRootMailReadyServiceRoleDefaultPolicy9A1F5C41",
        "Roles": [
          {
            "Ref": "testRootmailRootMailReadyServiceRole28CD8F41",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testRootmailRootMailReadyTriggerE528F2B1": {
      "DependsOn": [
        "testRootmailRootMailReadyTriggerServiceRoleE74AC2A4",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-1234-us-east-1",
          "S3Key": "3d21154d1ea7016302d23f7fac3b4baf934e92a8ca73c70bb807a280b17e3ff4.zip",
        },
        "Environment": {
          "Variables": {
            "SIGNAL_URL": {
              "Ref": "testRootmailRootMailReadyHandleC6B1EB65",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "testRootmailRootMailReadyTriggerServiceRoleE74AC2A4",
            "Arn",
          ],
        },
        "Runtime": "python3.9",
        "Timeout": 10,
      },
      "Type": "AWS::Lambda::Function",
    },
    "testRootmailRootMailReadyTriggerEventPattern0C7DAE47": {
      "Properties": {
        "EventPattern": {
          "detail": {
            "alarmName": [
              {
                "Ref": "testRootmailErrors174176D2",
              },
            ],
            "state": {
              "value": [
                "OK",
              ],
            },
          },
          "detail-type": [
            "CloudWatch Alarm State Change",
          ],
          "source": [
            "aws.cloudwatch",
          ],
        },
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "testRootmailRootMailReadyTriggerE528F2B1",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "testRootmailRootMailReadyTriggerEventPatternAllowEventRuletestStacktestRootmailRootMailReadyTriggerC42CFCEB14537371": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "testRootmailRootMailReadyTriggerE528F2B1",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "testRootmailRootMailReadyTriggerEventPattern0C7DAE47",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "testRootmailRootMailReadyTriggerServiceRoleE74AC2A4": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
